### PR TITLE
CSE Machine: Avoid pushing unnecessary env instructions (exhaustive version)

### PR DESCRIPTION
**Pull request by DiligentPenguinn**
### Description
This PR aims to implement an exhaustive version of PR #1699 and #1687 .
A `ControlItem` is environment-dependent if its evaluation depends on the current environment, OR its evaluation introduces environment-dependent items onto the control.
At each step, if the current control is free of environment-dependent items then we can avoid pushing environment instruction.
### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code quality improvements

### Some of my concerns
1) How exhaustive should the logic be? I think it’s enough to cover all node types and instructions allowed to be used in Source. Any remaining node types can default to being environment-independent

2) I initially wanted to place the logic in the AST file along with statementSeqTransform but I realized the logic needed to work for control items which can be either nodes or instructions. Instructions are only defined within the CSE Machine context so currently I place the logic there. 

3) I modified the test cases to use `generateCSEMachineStateStream` which allows me to check the number of environment-dependent items at each step, but now the test runs synchronously (because `generateCSEMachineStateStream` runs synchronously). Would this be a problem?

### CODE REVIEW

File: src/cse-machine/interpreter.ts [hunk 1]
```diff
@@ -107,12 +107,17 @@ export class Control extends Stack<ControlItem> {
   }
 
   public canAvoidEnvInstr(): boolean {
-    return this.numEnvDependentItems === 0
+    return this.numEnvDependentItems == 0
```
Comment (RichDom2185) [inline src/cse-machine/interpreter.ts:110]: Why was this changed from `===` to `==`?

File: src/cse-machine/interpreter.ts [hunk 2]
```diff
@@ -107,12 +107,17 @@ export class Control extends Stack<ControlItem> {
...
+
+  // For testing purposes
+  public getNumEnvDependentItems(): number {
+    return this.numEnvDependentItems
   }
 
   public pop(): ControlItem | undefined {
     const item = super.pop()
-    if (item !== undefined && isEnvDependent(item)) {
+    if (item != undefined && isEnvDependent(item)) {
```
Comment (RichDom2185) [inline src/cse-machine/interpreter.ts:120]: Was this change intentional? Note that `!= undefined` matches both (non-`null` and non-`undefined`) but `!== undefined` matches non-`undefined` only, i.e.

```ts
null == undefined; // true
null === undefined; // false
```

**Review by RichDom2185 — COMMENTED** [review]
Hi, did a quick look and just some questions below

File: src/cse-machine/interpreter.ts (hunk 1 above)
Comment (DiligentPenguinn) [reply src/cse-machine/interpreter.ts:110]: That was unintentional, I'll change it back
